stdlib/iarray.ml

```diff
@@ -0,0 +1,339 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*               Antal Spector-Zabusky, Jane Street, New York             *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*   Copyright 2023 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+open! Stdlib
+
+(* In this file, we use four different implementation strategies:
+
+     1. Reusing [external]s for mutable arrays.  (E.g., [get].)
+     2. Copying implementations from [array.ml], which in this new context read
+        from immutable arrays.  (E.g., [iter].)
+     3. As (2), but they construct a mutable array, which we unsafely
+        reinterpret as an immutable array at the very end (E.g., [map].)
+     4. (Only for sorting.) Copying the immutable array and passing it to a
+        mutating function.  (E.g., [sort].)
+
+   The first and third strategies are safe because mutable and immutable arrays
+   have the same runtime representation, and we only apply them to functions
+   that satisfy the following three properties:
+
+     A. They do not mutate their array inputs;
+     B. They do not hold on to their array inputs; and
+     C. They return a fresh array if they return an array.
+
+   We do not expose other primitives (e.g., [unsafe_set]) or functions (e.g.,
+   [fill]).
+
+   We choose between the four strategies as follows:
+
+     1. We use [external]s if there is a corresponding [external].
+     2. Functions that only read arrays can have their implementations safely
+        copied.
+     3. Functions that create an immutable array have to do so by creating a
+        mutable array and freezing it, since immutable arrays are, well,
+        immutable.  For efficiency, we freeze them unsafely.
+     4. Sorting is implemented in-place and this allows to to reuse it.
+*)
+
+[@@@ocaml.flambda_o3]
```

Comment:
This attribute does not exist in the mainline compiler, isn't it?

Reply:
No, it doesn’t seem to.

Reply:
Yes, the flambda-backend compiler recognizes it but upstream doesn't. I believe we do compile the standard library with `-O3` anyway, so this shouldn't make any difference even if we recognized it.